File: mission-to-mars-solid/src/main/java/mars/Helicoptere.java
package mars;

import java.util.stream.IntStream;

public class Helicoptere extends Rover implements ModuleMartien3D {

	public Helicoptere(Direction direction, Position position) {
		super(direction, position);
	}

	@Override
	public void monter() {
		position = new Position(position.getX(), position.getX(), position.getZ() + 1);
	}

	@Override
	public void descendre() {
		position = new Position(position.getX(), position.getX(), position.getZ() - 1);
	}

	@Override
	public void activerRecuperation() {
		atterir();
		super.activerRecuperation();
	}
	
	private void atterir() {
		IntStream.range(0, getPosition().getZ())
			.forEach((i -> descendre()));
	}
}

File: typescript/src/Helicoptere.ts
import { Direction } from './Direction'
import { Module } from './Module'
import { ModuleMartien3D } from './ModuleMartien3D'
import { Position } from './Position'

export class Helicoptere extends Module implements ModuleMartien3D {
  constructor(readonly direction: Direction, readonly position: Position) {
    super();
    super._direction = direction;
    super._position = position;
  }

  monter() {
    super._position = new Position(this._position.x, this._position.y, this._position.z + 1);
  }

  descendre() {
    super._position = new Position(this._position.x, this._position.y, this._position.z - 1);
  }

  override activerRecuperation() {
    this.atterir();
    super.activerRecuperation();
  }

  private atterir() {
    [...Array(this.getPosition().z)]
      .forEach(() => { this.descendre(); });
  }
}
